src/mod_extdisco.erl

```diff
@@ -0,0 +1,132 @@
+%%==============================================================================
+%% Copyright 2020 Erlang Solutions Ltd.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%% http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%==============================================================================
+-module (mod_extdisco).
+-author('jan.ciesla@erlang-solutions.com').
+
+-xep([{xep, 215}, {version, "0.7"}]).
+-behaviour(gen_mod).
+
+%% gen_mod callbacks.
+-export([start/2, stop/1]).
+
+-export([process_iq/4]).
+
+-include("mongoose.hrl").
+-include("jlib.hrl").
+
+-define(EXT_DISCO, <<"urn:xmpp:extdisco:2">>).
+
+-spec start(jid:server(), list()) -> ok.
+start(Host, Opts) ->
+    IQDisc = gen_mod:get_opt(iqdisc, Opts, one_queue),
+    mod_disco:register_feature(Host, ?EXT_DISCO),
+    gen_iq_handler:add_iq_handler(ejabberd_local, Host,
+        ?EXT_DISCO, ?MODULE, process_iq, IQDisc).
+
+-spec stop(jid:server()) -> ok.
+stop(Host) ->
+    mod_disco:unregister_feature(Host, ?EXT_DISCO),
+    gen_iq_handler:remove_iq_handler(ejabberd_sm, Host, ?EXT_DISCO).
+
+-spec process_iq(jid:jid(), jid:jid(), mongoose_acc:t(), jlib:iq()) ->
+    {mongoose_acc:t(), jlib:iq()}.
+process_iq(_From, _To, Acc, #iq{type = get, sub_el = SubEl} = IQ) ->
+    Response = case request_type(SubEl) of
+        all_services ->
+            RequestedServices = get_external_services(),
+            create_iq_response(RequestedServices);
+        {credentials, {Type, Host}} ->
+            Services = get_external_services([Type], []),
+            RequestedServices = lists:filter(fun({_, Opts}) ->
+                gen_mod:get_opt(host, Opts, undefined) == binary_to_list(Host)
+                end, Services),
+            create_iq_response_credentials(RequestedServices);
+        {selected_services, Type} ->
+            RequestedServices = get_external_services([Type], []),
+            create_iq_response_services(RequestedServices, Type);
+        _ ->
+            []
+    end,
+    {Acc, IQ#iq{type = result, sub_el = Response}}.
+
+request_type(#xmlel{name = <<"services">>} = Element) ->
+    case exml_query:attr(Element, <<"type">>) of
+        undefined -> all_services;
+        Services -> {selected_services, binary_to_atom(Services, utf8)}
```

Comment:
I would call this `Type`. It confused me for a while, thinking that `binary_to_atom(Services)`, being `Services` a plural, would fail as we're passing a list. Also, we just queried for `<<"type">>` 😄

Reply:
And also, does it make sense to use `binary_to_existing_atom`, or other kind of pattern-matching for allowed atoms? This is straight input from clients, so if sufficient atoms are given, this could kill a node.

Reply:
Good point! True, type makes more sense but actually going after pattern matching would make this safer.